### ISSUE TITLE
stage3 images do not have /run/lock directory required by some ebuilds

### ISSUE THREAD

**Issue by Leo3418** [gentoo/gentoo-docker-images#115]
`app-text/build-docbook-catalog-2.1` has been recently stabilized for multiple architectures and is now being pre-installed in stage3.  This version requires `/run/lock` to exist in the file system. If it does not exist, the `build-docbook-catalog` program gives the following error:

```console
# build-docbook-catalog 
build-docbook-catalog: //run/lock: missing critical system path; please create it, aborting
```

However, `/run/lock` is not present in Docker container created from the latest stage3 Docker images. This prevents other ebuilds that depend on this package from successfully being merged in the container, such as `app-text/docbook-xml-dtd-4.5-r2`.

To reproduce:

1. Get the latest version of the `gentoo/stage3` image, then start a container with interactive session:
   ```console
   # docker pull gentoo/stage3
   # docker run -it gentoo/stage3
   ```
2. Install an affected package in the container:
   ```console
   # emerge-webrsync 
   # emerge app-text/docbook-xml-dtd
   ```

<details>
  <summary>Execution log</summary>
  
```console
# docker pull gentoo/stage3
Trying to pull docker.io/gentoo/stage3:latest...
Getting image source signatures
Copying blob ea4b30c8b584 skipped: already exists  
Copying config c3cf53e81d done  
Writing manifest to image destination
Storing signatures
c3cf53e81d255ff0873fdfc7beadf9970ffcaf0f948befba223cd9d5af71297a
# docker run -it gentoo/stage3

# emerge-webrsync 
!!! Section 'gentoo' in repos.conf has location attribute set to nonexistent directory: '/var/db/repos/gentoo'
!!! Invalid Repository Location (not a dir): '/var/db/repos/gentoo'
Fetching most recent snapshot ...
Trying to retrieve 20211221 snapshot from http://distfiles.gentoo.org ...
Fetching file gentoo-20211221.tar.xz.md5sum ...
Fetching file gentoo-20211221.tar.xz.gpgsig ...
Fetching file gentoo-20211221.tar.xz ...
Checking digest ...
Getting snapshot timestamp ...
Syncing local tree ...

Number of files: 148,102 (reg: 121,273, dir: 26,829)
Number of created files: 148,101 (reg: 121,273, dir: 26,828)
Number of deleted files: 0
Number of regular files transferred: 121,273
Total file size: 236.51M bytes
Total transferred file size: 236.51M bytes
Literal data: 236.51M bytes
Matched data: 0 bytes
File list size: 2.63M
File list generation time: 0.001 seconds
File list transfer time: 0.000 seconds
Total bytes sent: 114.14M
Total bytes received: 2.42M

sent 114.14M bytes  received 2.42M bytes  11.10M bytes/sec
total size is 236.51M  speedup is 2.03
Cleaning up ...

Performing Global Updates
(Could take a couple of minutes if you have a lot of binary packages.)



 * IMPORTANT: 12 news items need reading for repository 'gentoo'.
 * Use eselect news read to view new items.

# emerge app-text/docbook-xml-dtd

 * IMPORTANT: 12 news items need reading for repository 'gentoo'.
 * Use eselect news read to view new items.

Calculating dependencies... done!

>>> Verifying ebuild manifests

>>> Emerging (1 of 1) app-text/docbook-xml-dtd-4.5-r2::gentoo
Unable to unshare: EPERM (for FEATURES="ipc-sandbox network-sandbox pid-sandbox")
bash: line 1: /var/cache/distfiles/.__portage_test_write__: Permission denied
Adjusting permissions recursively: '/var/cache/distfiles'
>>> Downloading 'http://distfiles.gentoo.org/distfiles/layout.conf'
--2021-12-22 01:13:05--  http://distfiles.gentoo.org/distfiles/layout.conf
Resolving distfiles.gentoo.org... 89.187.187.15, 2a02:6ea0:c800::8
Connecting to distfiles.gentoo.org|89.187.187.15|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 45 [text/plain]
Saving to: ‘/var/cache/distfiles/.layout.conf.distfiles.gentoo.org.__download__’

/var/cache/distfile 100%[===================>]      45  --.-KB/s    in 0s      

2021-12-22 01:13:05 (6.32 MB/s) - ‘/var/cache/distfiles/.layout.conf.distfiles.gentoo.org.__download__’ saved [45/45]

>>> Downloading 'http://distfiles.gentoo.org/distfiles/9e/docbook-xml-4.5.zip'
--2021-12-22 01:13:05--  http://distfiles.gentoo.org/distfiles/9e/docbook-xml-4.5.zip
Resolving distfiles.gentoo.org... 89.187.187.15, 2a02:6ea0:c800::8
Connecting to distfiles.gentoo.org|89.187.187.15|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 98497 (96K) [application/zip]
Saving to: ‘/var/cache/distfiles/docbook-xml-4.5.zip.__download__’

/var/cache/distfile 100%[===================>]  96.19K  --.-KB/s    in 0.008s  

2021-12-22 01:13:05 (12.1 MB/s) - ‘/var/cache/distfiles/docbook-xml-4.5.zip.__download__’ saved [98497/98497]

 * docbook-xml-4.5.zip BLAKE2B SHA512 size ;-) ...                       [ ok ]
Unable to unshare: EPERM (for FEATURES="pid-sandbox")
Unable to unshare: EPERM (for FEATURES="ipc-sandbox network-sandbox pid-sandbox")
>>> Unpacking source...
>>> Unpacking docbook-xml-4.5.zip to /var/tmp/portage/app-text/docbook-xml-dtd-4.5-r2/work
>>> Source unpacked in /var/tmp/portage/app-text/docbook-xml-dtd-4.5-r2/work
Unable to unshare: EPERM (for FEATURES="ipc-sandbox network-sandbox pid-sandbox")
>>> Preparing source in /var/tmp/portage/app-text/docbook-xml-dtd-4.5-r2/work ...
>>> Source prepared.
Unable to unshare: EPERM (for FEATURES="ipc-sandbox network-sandbox pid-sandbox")
>>> Configuring source in /var/tmp/portage/app-text/docbook-xml-dtd-4.5-r2/work ...
>>> Source configured.
Unable to unshare: EPERM (for FEATURES="ipc-sandbox network-sandbox pid-sandbox")
>>> Compiling source in /var/tmp/portage/app-text/docbook-xml-dtd-4.5-r2/work ...
>>> Source compiled.
Unable to unshare: EPERM (for FEATURES="ipc-sandbox network-sandbox pid-sandbox")
>>> Test phase [not enabled]: app-text/docbook-xml-dtd-4.5-r2
Unable to unshare: EPERM (for FEATURES="ipc-sandbox network-sandbox pid-sandbox")

>>> Install app-text/docbook-xml-dtd-4.5-r2 into /var/tmp/portage/app-text/docbook-xml-dtd-4.5-r2/image
>>> Completed installing app-text/docbook-xml-dtd-4.5-r2 into /var/tmp/portage/app-text/docbook-xml-dtd-4.5-r2/image

 * Final size of build directory: 536 KiB
 * Final size of installed tree:  536 KiB


>>> Installing (1 of 1) app-text/docbook-xml-dtd-4.5-r2::gentoo
Unable to unshare: EPERM (for FEATURES="ipc-sandbox network-sandbox pid-sandbox")
build-docbook-catalog: //run/lock: missing critical system path; please create it, aborting
 * Updating /etc/sgml/catalog ...                                        [ ok ]
Unable to unshare: EPERM (for FEATURES="ipc-sandbox network-sandbox pid-sandbox")
build-docbook-catalog: //run/lock: missing critical system path; please create it, aborting
 * ERROR: app-text/docbook-xml-dtd-4.5-r2::gentoo failed (postinst phase):
 *   Failed to regenerate docbook catalog. Is /run mounted?
 * 
 * Call stack:
 *     ebuild.sh, line 127:  Called pkg_postinst
 *   environment, line 448:  Called die
 * The specific snippet of code:
 *       build-docbook-catalog || die "Failed to regenerate docbook catalog. Is /run mounted?";
 * 
 * If you need support, post the output of `emerge --info '=app-text/docbook-xml-dtd-4.5-r2::gentoo'`,
 * the complete build log and the output of `emerge -pqv '=app-text/docbook-xml-dtd-4.5-r2::gentoo'`.
 * The complete build log is located at '/var/tmp/portage/app-text/docbook-xml-dtd-4.5-r2/temp/build.log'.
 * The ebuild environment file is located at '/var/tmp/portage/app-text/docbook-xml-dtd-4.5-r2/temp/environment'.
 * Working directory: '/var/tmp/portage/app-text/docbook-xml-dtd-4.5-r2/empty'
 * S: '/var/tmp/portage/app-text/docbook-xml-dtd-4.5-r2/work'
 * FAILED postinst: 1
Unable to unshare: EPERM (for FEATURES="ipc-sandbox network-sandbox pid-sandbox")

>>> Recording app-text/docbook-xml-dtd in "world" favorites file...
```

</details>

Should `/run/lock` now be created in the stage3 Docker images, so ebuilds like `app-text/docbook-xml-dtd` can be merged directly without any issues out-of-box?

**Comment by MeisterP** [comment]
Hitting the same issue.
The changes to the docbook ebuilds were introduced in  https://gitweb.gentoo.org/repo/gentoo.git/commit/?id=26a4a7456ad6509d42ef1c3f92aee8eea3f7a04b because of https://bugs.gentoo.org/816303

**Comment by KSmanis** [comment]
> Should `/run/lock` now be created in the stage3 Docker images, so ebuilds like `app-text/docbook-xml-dtd` can be merged directly without any issues out-of-box?

I don't think so, containers are not fully functioning installations. `/run` is a detail handled by the init system, which I would say is out of our scope.

**Comment by Leo3418** [comment]
> I don't think so, containers are not fully functioning installations. `/run` is a detail handled by the init system, which I would say is out of our scope.

Well, a container usually does not use an init system at all, so there is really no program that can handle creation of `/run/lock` during the runtime. Using something like `-v /run:/run` could do the job, but it would unnecessarily expose part of the host system's state to the container, which is not always desirable.

As a stage3 image user, I'd say that creating an empty `/run/lock` in the image would improve user experience. This only requires an additional `mkdir -p /run/lock` invocation in the Dockerfile, like https://github.com/Leo3418/gentoo-java-image/commit/4b097d38bc6bd32e42b5d7ce6a5fcad165251862, so it's very easy to implement. Yet it can help reduce the number of support requests in the future when another user attempts to do the same thing as me and @MeisterP have done.

To give a little bit of the context, I'm using containers created from the `gentoo/stage3` image to do ebuild testing. Gentoo Wiki has [a page](https://wiki.gentoo.org/wiki/Test_environment) that briefly describes how Docker containers can be used for ebuild testing, and there's a `dev-util/ebuildtester` utility that allows users to test ebuilds with this image, so I think there are a non-trivial amount of people who are also doing their ebuild tests in containers, even though "containers are not fully functioning installations". And when ebuilds are installed in a container for testing, dependencies like `app-text/docbook-xml-dtd` might be pulled, and the testers would run into the same issue, unless they had created `/run/lock` themselves in advance. But, why don't we just do that favor for them in the Dockerfile?

**Comment by gyakovlev** [comment]
recent versions of docbook were modified to use /etc/xml instead of /run

should not be a problem with latest images anymore.

**Comment by gyakovlev** [comment]
https://gitweb.gentoo.org/proj/build-docbook-catalog.git/commit/?id=301eedd615b1f9f7369f39c06831dbec84235c08
https://github.com/gentoo/gentoo/commit/9048b7e5e28d3e81db2043443c6f92ca718ccbfa

relevant commits.
basically releng had similar problem building stage3 files

**Comment by Leo3418** [comment]
> recent versions of docbook were modified to use /etc/xml instead of /run/lock
> 
> should not be a problem with latest images anymore.

Right, I just tried in a container created from the latest image to upgrade `app-text/build-docbook-catalog` to 2.2 and emerge `app-text/docbook-xml-dtd`, and the error would no longer show up. `build-docbook-catalog` 2.1 has already been removed from `::gentoo`. Hopefully 2.2 can get stabilized soon, but with 1.21 -- the only stabilized version and the version included in stage3 as of now -- there is no such issue either. Thus, this issue should not affect the latest images anymore.

Thank you folks for the work!!
